### ISSUE TITLE
Support Apache Camel

### ISSUE THREAD

**Issue by jkschneider** [micrometer-metrics/micrometer#410]

**Comment by bjartek** [comment]
I have to look internally at our Camel stuff but could this be as «simple» as Creating some producers? 
http://camel.apache.org/metrics-component.html

How do we handle labling? We have to parse the to string somehow?

**Comment by jacovt** [comment]
As I see it, there might be 2 parts to this (unless I am completely missing the boat):

1. Implement it into the metrics component as a producer.
2. Create a MeterBinder implementation that simply exposes all the camel metrics that are already exists in JMX. Something similar to what is being done for Tomcat: 
[https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/tomcat/TomcatMetrics.java]

I think we might have to do a little fancy naming footwork with the JMX metrics to make it easier to graph in something like Grafana.

Thoughts?

Thanks,
---Jaco

**Comment by jkschneider** [comment]
@ohr Biggest question you raise is also a question of mine: where should this live?

The best possible scenario (for us) is that instrumentation lives in Camel itself (similar to what was done for rabbitmq-java and hikaricp), but I know that isn't always possible. The binders we've included in the core module so far are chosen deliberately for their long-established API stability. 

We could of course create a new module for each new integration, `micrometer-httpclient`, `micrometer-okhttp`, etc. And of course Prometheus has decided to do this with `simpleclient_xxx` for lots of different integrations. Developer experience is kind of rough though: first there is the discoverability problem of just knowing what bindings exist, and then there is the problem of including additional dependencies and keeping their versions aligned. I can see a future where we have `io.micrometer.contrib:micrometer-camel` or something like that if we want binders for products that have less stable APIs (Camel may or may not, I don't know either way). We just haven't encountered that yet. The line may always be a bit blurry.

**Comment by jkschneider** [comment]
`MetricsRoutePolicy` is already directly tied to Dropwizard Metrics. We could certainly build timers off of the `onExchangeBegin`/`onExchangeDone` events, but Camel already appears to have a bit of a leaky opinion here.

**Comment by checketts** [comment]
@bjartek Jon did mention that option when he referred to the Prometheus `simpleclient_xxx` layout.  What I've seen in that project is that after a while they stopped accepting additional modules.

The only problem I see with leaving them in core is the need to be careful that the camel dep remains `optional` and class usage is kept within that binder.

Is anyone here a camel contributor that could try putting the micrometer support directly in Camel?

**Comment by ohr** [comment]
I contributed a couple of things to Camel already so I could prepare a pull request in Camel for micrometer support (following the pattern from camel-metrics. @davsclaus: any objections?)

Note, however, that Camel has already a LOT of components that integrate 3rd party libs, and a considerable percentage of them have a tendency to become outdated over time (i.e. not catching up with new versions and features of the 3rd party libs being covered). 
So, depending on further development here (particularly when introducing breaking changes!), this would need ongoing attention.

**Comment by jkschneider** [comment]
@ohr Being a Pivotal sponsored project, Micrometer is bound to 3 year support for major releases (no breaking public API changes inside a major release) and 1 year for minor.

**Comment by davsclaus** [comment]
@ohr yeah that would be lovely. There is a JIRA ticket already: https://issues.apache.org/jira/browse/CAMEL-11600

I disagree a bit on your about components gets outdated, we upgrade to latest versions as much as possible, and include all PRs that people contribute, and we overhaul and redo components on major new versions like elasticsearch. The camel-spark component however indeed needs an upgrade to v2 or to do a camel-spark2 component. Use the JIRA voting to help prioritize and also contribute as you have done.

**Comment by jonmcewen** [comment]
@jkschneider great.  Is there any further documentation or example code for this?

**Comment by ohr** [comment]
A Camel Dropwizard gauge maintains its "private" value that is actively set when an exchange passes the producer:
```
   Object subject = exchange.getIn().getHeader(HEADER_GAUGE_SUBJECT, Object.class);
   if (subject != null) {
      camelMetricsGauge.setValue(subject);
   }
```
Micrometer observes a weak reference, so I guess that simply having the Camel route to modify the target of that reference is preferrable.

With your remarks, I skip JMX stuff for the time being.
In fact I wrote a Jackson-based JSON serializer (following the pattern of the camel-metrics module), so I keep it over there.

**Comment by ohr** [comment]
CAMEL-11600 was resolved. So this could bei closed as well.

**Comment by shakuzen** [comment]
As mentioned by @izeye and @ohr, Micrometer support is available in Apache Camel since Apache Camel 2.22.0
